pymbar/mbar_solvers.py

```diff
@@ -289,105 +297,87 @@ def adaptive(u_kn, N_k, f_k, tol=1.0e-12, options=None):
-                else:
-                    logger.info(
-                        "Choosing self-consistent iteration for lower gradient on iteration %d"
-                        % iteration
-                    )
+            which_method.append("sci")
         else:
             f_k = f_nr
             g = g_nr
             nr_iter += 1
-            if options["verbose"]:
-                logger.info("Newton-Raphson used on iteration %d" % iteration)
-
-        div = np.abs(f_k[1:])  # what we will divide by to get relative difference
-        zeroed = np.abs(f_k[1:]) < np.min(
-            [10 ** -8, tol]
-        )  # check which values are near enough to zero, hard coded max for now.
-        div[zeroed] = 1.0  # for these values, use absolute values.
-        max_delta = np.max(np.abs(f_k[1:] - f_old[1:]) / div)
-        if np.isnan(max_delta) or (max_delta < tol):
-            doneIterating = True
+            which_method.append("nr")
+            
+        div = jnp.abs(f_k[1:]) # what we will divide by to get relative difference
+        zeroed = jnp.abs(f_k[1:])< np.min([10**-8,tol]) # check which values are near enough to zero, hard coded max for now.
+        jax.ops.index_update(div,index[zeroed],1.0)
+        max_delta = jnp.max(jnp.abs(f_k[1:]-f_old[1:])/div)
+        if jnp.isnan(max_delta) or (max_delta < tol):
             break
 
-    if doneIterating:
-        if options["verbose"]:
-            logger.info(
-                "Converged to tolerance of {:e} in {:d} iterations.".format(
-                    max_delta, iteration + 1
-                )
-            )
-            logger.info(
-                "Of {:d} iterations, {:d} were Newton-Raphson iterations and {:d} were self-consistent iterations".format(
-                    iteration + 1, nr_iter, sci_iter
-                )
-            )
-            if np.all(f_k == 0.0):
-                # all f_k appear to be zero
-                logger.info("WARNING: All f_k appear to be zero.")
-    else:
-        logger.warning("WARNING: Did not converge to within specified tolerance.")
-        if options["maximum_iterations"] <= 0:
-            logger.warning(
-                "No iterations ran be cause maximum_iterations was <= 0 ({:s})!".format(
-                    options["maximum_iterations"]
-                )
-            )
-        else:
-            logger.warning(
-                "max_delta = {:e}, tol = {:e}, maximum_iterations = {:d}, iterations completed = {:d}".format(
-                    max_delta, tol, options["maximum_iterations"], iteration
-                )
-            )
-    return f_k
+    results = dict()
+    results['f_k'] = f_k
+    results['max_delta'] = max_delta
+    results['iterations'] = iteration
+    results['gnorms_sci'] = gnorms_sci
+    results['gnorms_nr'] = gnorms_nr
+    results['which_method'] = which_method
+
+    return results
 
+jit_core_adaptive = jax.jit(core_adaptive,static_argnums=(0,1,2,3,4,5,))
```

Comment:
When you call jax.jit, the first invocation will be exceedingly slow it's compiling the JIT kernels. So if you're benchmarking you should call this *more* than once to amortize out the jit time.

Futhermore, if the arguments have shapes that contain (None,) in one of the shapes, a recompilation is required due to an XLA requirement for each specialization of (None,) into a known shape.

Reply:
Ah, this explains why the partial jitification works better- then it's calling different functions multiple times.  Though only 7-8 times in many cases.

Reply:
I will test running mbar initialization several times with different data sets in the same script to see if that helps.

Reply:
I also noticed you're declaring every arg to be static, which means that unless you're passing in the exact args you'll be triggering a recompilation. Why did you have to declare them to be static to start with?

Reply:
U_kn probably shouldn't be static if you're trying to pass in results from different runs?

Reply:
I had to declare it static because otherwise, gnorm_sci and gnorm_nr end up as traced arrays instead of concrete arrays (since they are functions of traced arrays), and jit refuses to compile the conditional comparing the two, which decides the branch to take, unless they are declared static.

Reply:
@proteneer had a good suggestion for handling the conditional:

import jax
import jax.numpy as jnp
# can't be JIT'd
def foo(a, b, cond_a, cond_b):
    if cond_a < cond_b:
        return a+b
    else:
        return a-b
# foo_jit = jax.jit(foo) # fails
print(foo(0., 1., 2., 3.))
# can be JIT'd
def bar(a, b, cond_a, cond_b):
    return jnp.where(cond_a < cond_b, a+b, a-b)
bar_jit = jax.jit(bar)
print(bar_jit(0., 1., 2., 3.))

Reply:
So, if we accelerate just the inner loop, then we can pull the conditional out of the accelerated code with little loss of timing - all of the conditionals are just comparing floats and assigning values, so there is little use in accelerating them.